torch/utils/module_tracker.py

```diff
@@ -74,6 +93,54 @@ def is_bw(self):
         """
         return torch._C._current_graph_task_id() != -1
 
+    def get_known_fqn(self, mod):
+        """
+        Return the fqn for the given module if it is known to the Module Tracker, otherwise None
+        """
+
+        if mod not in self._known_modules:
+            return None
+        return self._known_modules[mod]
+
+    def register_user_hooks(
+        self,
+        pre_fw_hook: Optional[Callable] = None,
+        post_fw_hook: Optional[Callable] = None,
+        pre_bw_hook: Optional[Callable] = None,
+        post_bw_hook: Optional[Callable] = None,
+    ):
+        """
+        Registers user specified hooks that are called before/after the forward/backward pass for
+        each module tracked by the Module Tracker. One or more can be none.
+
+        The hooks should have the following signatures.
+
+        pre_fw_hook (fqn, module, input) : Called before the forward pass for the module.
+        post_fw_hook (fqn, module, input, output) : Called after the forward pass for the module.
+        pre_bw_hook (fqn, *outputs) : A multi-grad hook that is called on all the outputs of the
+                                      module that require gradients.
+        post_bw_hook (fqn, *inputs) : A multi-grad hook that is called on all the inputs of the
+                                      module that require gradients.
+
+        They have the same signature, constraints and requirements as the hooks registered with
+        ``register_module_forward_pre_hook``, ``register_module_forward_hook`` and
+        ``register_multi_grad_hook`` with an additional fqn arg that is supplied by the Module Tracker.
+        The fqn will be present in the ``parents`` attribute when each these hooks is called.
+        """
+        self._user_pre_fw_hook = pre_fw_hook
```

Comment:
We should error if this is about to silently override existing hooks.

Reply:
No we are not overriding them, but just using them as markers as you said.